design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,219 @@
+After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non Goals
+Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
+* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadCompleteToEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - EmulatorThreadCompleteToEvenParity
+```
+This annotation will be respected only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.
+
+## Virt-API:
+
+### VMI mutator webhook
+When the following conditions are met:
+1. The feature gate is enabled on the Kubevirt CR
+2. Kubevirt CR has the annotation.
+3. VMI is configured with `isolateEmulatorThread: true`
+
+then the VMI mutator webhook will copy the annotation from the kubevirt CR to the VMI annotations. This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.
+
+### VMI create-admitter validation webhook
+Should allow the VMI to have the new annotation only if:
+1. The feature gate is enabled on the Kubevirt CR
```

Comment:
We want users to control this feature on the cluster-level through the CR. We should not care if they set it on VM-level using an undocumented annotation. Consider removing this check together with the gate.

Reply:
I tend to agree.

Reply:
@phoracek please note that with the current design, removing this check will bypass the feature gate on the VMI level - and the behavior change caused by the annotation will take affect.

Reply:
I'm for *having* a feature gate controlling the new functionality. But we should not reject a VM with an annotation just because a FG is not set. Today if I set `EmulatorThreadCompleteToEvenParity=true` on my VM it is admitted to the system. This should should not change in the future. Admission webhooks should reject VMs with invalid fields (e.g a feature-gated field when the gate is closed). But annotations are lightweight. They are not subject to the same structure scrutiny as actual fields. Typically, any RC owner can create any annotation on their RC without risk of rejection.

Reply:
Annotations with kubevirt subdomain shouldn't be used outside of kubevirt. What would be the behavior if we do not reject the annotated VM when the FG is not set? Would you want an FG check all over the codebase? I don't want it because the result of such a check could change over time and we could quickly get into a weird state. Therefore if a VM has this annotation it should behave as described even if the FG is off in the future.

Reply:
> What would be the behavior if we do not reject the annotated VM when the FG is not set?

Undefined. This would require us to remove the documentation of the per-VM annotation from https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#requesting-dedicated-cpu-for-qemu-emulator. Then whoever would use the per-VM annotation would be doing it on their own risk.

With this we would have:
* One less feature gate
* One less VM-level knob
* One less webhook check

>  Would you want an FG check all over the codebase?

No. What I suggested is not having the VM-level annotation checked against the feature gate at all.

Reply:
The annotation is not released. The documentation is lying. Nothing is holding us to remove the per-VM documentation if we decide we want to support only cluster-wide knob.

>Undefined.

Doesn't sound as safe and sane as I would like. Even if this is experimental, the behavior or feature overall should be complete in some sense (even in a minimalistic sense) for users to provide at least some relevant feedback.

Again, we have feature gates and we described their meaning well today. I don't see any reason for not using them. The less, less, less argument is not appealing enough for me if in the end, we will introduce the feature gate eventually.

>>   Would you want an FG check all over the codebase?
>No. What I suggested is not having the VM-level annotation checked against the feature gate at all.

So only the presence of the annotation would decide the semantic? That would be fine but again, we have never defined the lifecycle of annotations and we used feature gates for everything else. I don't see a compelling reason to diverge now.

Reply:
I think you got my point, and although I disagree, I won't block this PR anymore if you insist on using webhooks and feature gates here.

Reply:
@phoracek just to be sure, we can do this in the future if we decide on this as a project. But otherwise, we should follow our own rules &/ consistent. We already have a few more examples of how to do one thing than we should have.

Reply:
> Would you want an FG check all over the codebase?

Yes. I think that it is important. If we decide that a gated feature is going to be dropped, we would like to delete all relevant code. Thus I think that the each dependent behavior should have have a very clear `if (FG==true)` clause.

Much like @phoracek I don't block on this. I just don't like the fact that it imposes non-eventually-consistent behavior.

Reply:
>Yes. I think that it is important. If we decide that a gated feature is going to be dropped, we would like to delete all relevant code. Thus I think that the each dependent behavior should have have a very clear if (FG==true) clause.

I disagree, not only does it complicate the code but it also increases the chance of the system converging to an undefined state. Once a VM with some feature is admitted into the system, the system should be able to handle it. Of course, if we are talking about alpha/beta or gated functionality, we might not guarantee that all cases are covered or for example, upgrades will not break the system.

Reply:
>I don't see what exactly the feature gate is protecting here. It's only the admin that can set the cluster-wide functionality anyway - using the CR annotation.
On a VM level, yes, the user can overallocate another CPU in a very specific case - I don't see an issue here.

What about accidental usage? (copy-pasting from the internet? vendor vs Kubevirt? and so on) I mean if I take these arguments I have a hard time figuring out why we have feature gates. Did we forget about why feature gates exist? For example as an admin, I would like to have control over deployed systems. Or I would like to know when I use something I should have not.

>We should introduce a feature gate only when the is an actual API. Then the feature gate will be checked in the virt-config, not in the virt-launcher.
As long as this functionality is using an experimental annotation I don't see a need for a feature gate.

If you think so, please let`s write this down for future guidance but otherwise, I can find you at least one counter-example.

>I wouldn't reject VMIs as well. The feature gate should control a specific functionality, in the worst case it shouldn't just allocate an additional CPU (that would lead to a cpu manager rejection anyway in the discussed case)

I apologize but I do not follow. The feature gate is controlling a specific functionality or not? I also don't understand the part about the worst case.

Reply:
> > I don't see what exactly the feature gate is protecting here. It's only the admin that can set the cluster-wide functionality anyway - using the CR annotation.
> > On a VM level, yes, the user can overallocate another CPU in a very specific case - I don't see an issue here.
> 
> What about accidental usage? (copy-pasting from the internet? vendor vs Kubevirt? and so on) I mean if I take these arguments I have a hard time figuring out why we have feature gates. Did we forget about why feature gates exist? For example as an admin, I would like to have control over deployed systems. Or I would like to know when I use something I should have not.
> 

In my view, feature gates are generally needed to protect against the use of unstable features / APIs or to enable/disable functionality on a cluster that doesn't have another control.
In this case, we don't have an API, on the other hand, this is not a complete feature either.  This is an extension to an existing logic of the `isolateEmulatorThread`.
By setting this special annotation the admin provides us with consent to enable this logic on a cluster level.

> > We should introduce a feature gate only when the is an actual API. Then the feature gate will be checked in the virt-config, not in the virt-launcher.
> > As long as this functionality is using an experimental annotation I don't see a need for a feature gate.
> 
> If you think so, please let`s write this down for future guidance but otherwise, I can find you at least one counter-example.

I'm sure there are more examples, but in this case, I don't think the feature gate provides any benefits.
From my pov, the feature gate could be incorporated into virt-config function, when we will have a formal API. 

> 
> > I wouldn't reject VMIs as well. The feature gate should control a specific functionality, in the worst case it shouldn't just allocate an additional CPU (that would lead to a cpu manager rejection anyway in the discussed case)
> 
> I apologize but I do not follow. The feature gate is controlling a specific functionality or not? I also don't understand the part about the worst case.

The current annotation extends an existing functionality. If set and other conditions are met, the VMI will allocate an additional CPU, so that the emulator thread will be bound to two cpus. 
If not set, one CPU will be allocated for the emulator thread. 

What I am trying to say is that even if we'll decide to keep the feature gate, it should affect the number of allocated CPUs and not prevent the creation of VMIs. 

I hope it makes sense.